daemon/cmd/policy.go

```diff
@@ -92,68 +71,125 @@ func newPolicyTrifecta(params policyParams) (policyOut, error) {
 	// Called for various events, such as named port changes
 	// or certain identity updates.
-	policyUpdater := policy.NewUpdater(repo, params.EndpointManager)
+	policyUpdater := policy.NewUpdater(params.PolicyRepository, params.EndpointManager)
+
+	params.Lifecycle.Append(cell.Hook{
+		OnStop: func(hc cell.HookContext) error {
+			policyUpdater.Shutdown()
+			return nil
+		},
+	})
+
+	return policyUpdater
+}
+
+type identityAllocatorParams struct {
+	cell.In
+
+	Lifecycle        cell.Lifecycle
+	PolicyRepository *policy.Repository
+	PolicyUpdater    *policy.Updater
+}
+
+type identityAllocatorOut struct {
+	cell.Out
 
+	IdentityAllocator      CachingIdentityAllocator
+	CacheIdentityAllocator cache.IdentityAllocator
+	RemoteIdentityWatcher  clustermesh.RemoteIdentityWatcher
+	IdentityObservable     stream.Observable[cache.IdentityChange]
+}
+
+func newIdentityAllocator(params identityAllocatorParams) identityAllocatorOut {
 	// iao: updates SelectorCache and regenerates endpoints when
 	// identity allocation / deallocation has occurred.
 	iao := &identityAllocatorOwner{
-		policy:        repo,
-		policyUpdater: policyUpdater,
+		policy:        params.PolicyRepository,
+		policyUpdater: params.PolicyUpdater,
 	}
 
 	// Allocator: allocates local and cluster-wide security identities.
 	idAlloc := cache.NewCachingIdentityAllocator(iao)
 	idAlloc.EnableCheckpointing()
 
+	params.Lifecycle.Append(cell.Hook{
+		OnStop: func(hc cell.HookContext) error {
+			idAlloc.Close()
+			return nil
+		},
+	})
+
+	return identityAllocatorOut{
+		IdentityAllocator:      idAlloc,
+		CacheIdentityAllocator: idAlloc,
+		RemoteIdentityWatcher:  idAlloc,
+		IdentityObservable:     idAlloc,
+	}
+}
+
+type ipCacheParams struct {
+	cell.In
+
+	Lifecycle              cell.Lifecycle
+	CacheIdentityAllocator cache.IdentityAllocator
+	PolicyRepository       *policy.Repository
+	EndpointManager        endpointmanager.EndpointManager
+	CacheStatus            synced.CacheStatus
+}
+
+func newIPCache(params ipCacheParams) *ipcache.IPCache {
+	ctx, cancel := context.WithCancel(context.Background())
+
 	// IPCache: aggregates node-local prefix labels and allocates
 	// local identities. Generates incremental updates, pushes
 	// to endpoints.
 	ipc := ipcache.NewIPCache(&ipcache.Configuration{
 		Context:           ctx,
-		IdentityAllocator: idAlloc,
-		PolicyHandler:     iao.policy.GetSelectorCache(),
+		IdentityAllocator: params.CacheIdentityAllocator,
+		PolicyHandler:     params.PolicyRepository.GetSelectorCache(),
 		DatapathHandler:   params.EndpointManager,
 		CacheStatus:       params.CacheStatus,
 	})
 
 	params.Lifecycle.Append(cell.Hook{
-		OnStart: func(hc cell.HookContext) error {
-			iao.policy.Start()
-			return nil
-		},
 		OnStop: func(hc cell.HookContext) error {
 			cancel()
 
-			// Preserve the order of shutdown but still propagate the error
-			// to hive.
-			err := ipc.Shutdown()
-			policyUpdater.Shutdown()
-			idAlloc.Close()
-
-			return err
```

Comment:
I think this change slightly changes the ordering. As far as I can tell, there is no dependency between idAlloc and policyUpdater and so the shutdown hooks might be executed in a different order now.

In my experience, the shutdown is not well tested in CI. I recommend doing a manual test to see if there are any logged errors during shutdown.

Reply:
Thanks for highlighting this @gandro. I had the same thoughts during development but thought that it's worth to start with the inverted shutdown order aligned with the Hive dependencies (without introducing "pseudo" dependencies).

I did a manual shutdown test by deleting an agent pod. And there are no errors in the log - actually there aren't any logs about shutting down at all (even on debug) :thinking: ) Any other (preferred) way to manually test the shutdown?

```
...
time="2024-08-13T09:14:23Z" level=debug msg="Starting new iteration of background sync" subsys=nodemanager syncInterval=1m5.91673732s
time="2024-08-13T09:14:25Z" level=debug msg="Pod is using host networking" k8sNamespace=kube-system k8sPodName=cilium-lxxlh subsys=k8s-watcher
time="2024-08-13T09:14:25Z" level=debug msg="Pod does not have any annotations nor labels changed" new-annotations="map[container.apparmor.security.beta.kubernetes.io/apply-sysctl-overwrites:unconfined container.apparmor.security.beta.kubernetes.io/cilium-agent:unconfined container.apparmor.security.beta.kubernetes.io/clean-cilium-state:unconfined container.apparmor.security.beta.kubernetes.io/config:unconfined container.apparmor.security.beta.kubernetes.io/install-cni-binaries:unconfined container.apparmor.security.beta.kubernetes.io/mount-bpf-fs:unconfined container.apparmor.security.beta.kubernetes.io/mount-cgroup:unconfined]" new-labels="map[app.kubernetes.io/name:cilium-agent app.kubernetes.io/part-of:cilium controller-revision-hash:66f578f6c6 k8s-app:cilium pod-template-generation:1]" old-annotations="map[container.apparmor.security.beta.kubernetes.io/apply-sysctl-overwrites:unconfined container.apparmor.security.beta.kubernetes.io/cilium-agent:unconfined container.apparmor.security.beta.kubernetes.io/clean-cilium-state:unconfined container.apparmor.security.beta.kubernetes.io/config:unconfined container.apparmor.security.beta.kubernetes.io/install-cni-binaries:unconfined container.apparmor.security.beta.kubernetes.io/mount-bpf-fs:unconfined container.apparmor.security.beta.kubernetes.io/mount-cgroup:unconfined]" old-labels="map[app.kubernetes.io/name:cilium-agent app.kubernetes.io/part-of:cilium controller-revision-hash:66f578f6c6 k8s-app:cilium pod-template-generation:1]" subsys=k8s-watcher
time="2024-08-13T09:14:25Z" level=debug msg="Processing 2 endpoints for EndpointSlice hubble-peer-sqpjb" subsys=k8s
time="2024-08-13T09:14:25Z" level=debug msg="Endpoint address 172.19.0.2 on EndpointSlice hubble-peer-sqpjb is Terminating" subsys=k8s
time="2024-08-13T09:14:25Z" level=debug msg="EndpointSlice hubble-peer-sqpjb has 2 backends" subsys=k8s
time="2024-08-13T09:14:25Z" level=debug msg="Processing 2 endpoints for EndpointSlice hubble-peer-sqpjb" subsys=k8s
time="2024-08-13T09:14:25Z" level=debug msg="Endpoint address 172.19.0.2 on EndpointSlice hubble-peer-sqpjb is Terminating" subsys=k8s
time="2024-08-13T09:14:25Z" level=debug msg="EndpointSlice hubble-peer-sqpjb has 2 backends" subsys=k8s
time="2024-08-13T09:14:25Z" level=debug msg="Kubernetes service definition changed" action=service-updated endpoints="172.19.0.2:4244/TCP,172.19.0.3:4244/TCP" k8sNamespace=kube-system k8sSvcName=hubble-peer old-endpoints="172.19.0.2:4244/TCP,172.19.0.3:4244/TCP" old-service=nil service="frontends:[10.96.161.224]/ports=[peer-service]/selector=map[k8s-app:cilium]" subsys=k8s-watcher
time="2024-08-13T09:14:25Z" level=debug msg="Upserting service" backends="[[172.19.0.2:4244,State:terminating] [172.19.0.3:4244,State:active]]" l7LBProxyPort=0 loadBalancerSourceRanges="[]" serviceIP="{10.96.161.224 {TCP 443} 0}" serviceName=hubble-peer serviceNamespace=kube-system sessionAffinity=false sessionAffinityTimeout=0 subsys=service svcExtTrafficPolicy=Cluster svcHealthCheckNodePort=0 svcIntTrafficPolicy=Local svcType=ClusterIP
time="2024-08-13T09:14:25Z" level=debug msg="Acquired service ID" backends="[[172.19.0.2:4244,State:terminating] [172.19.0.3:4244,State:active]]" l7LBProxyPort=0 loadBalancerSourceRanges="[]" serviceID=8 serviceIP="{10.96.161.224 {TCP 443} 0}" serviceName=hubble-peer serviceNamespace=kube-system sessionAffinity=false sessionAffinityTimeout=0 subsys=service svcExtTrafficPolicy=Cluster svcHealthCheckNodePort=0 svcIntTrafficPolicy=Local svcType=ClusterIP
time="2024-08-13T09:14:25Z" level=debug msg="Deleting backends from session affinity match" backends="[]" serviceID=8 subsys=service
time="2024-08-13T09:14:25Z" level=debug msg="Upserted service entry" backendSlot=1 subsys=map-lb svcKey="10.96.161.224:47873 (1)" svcVal="2 0[0] (2048) [0x0 0x0]"
time="2024-08-13T09:14:25Z" level=debug msg="Upserted service entry" backendSlot=0 subsys=map-lb svcKey="10.96.161.224:47873 (0)" svcVal="0 1[0] (2048) [0x0 0x10]"
time="2024-08-13T09:14:25Z" level=debug msg="Adding backends to affinity match map" backends="[]" serviceID=8 subsys=service

```

Reply:
`kubectl -n kube-system exec cilium-dqtt9 -- kill 1` does the trick for me

Reply:
ok even less logoutput with this :smiley: at least no error

Reply:
That is weird, you should see
```
time=2024-08-13T09:38:31Z level=info msg="Signal received" signal=terminated
time=2024-08-13T09:38:31Z level=info msg=Stopping
```

followed by debug statement for the stop hooks

Reply:
OK, problem (of not seeing the log) seems to be related to running Cilium in debug mode (`make kind-debug`). Probably an issue with the signal not being properly forwarded to the process. :facepalm: :shrug: 

But it works fine when using `make kind-image` & `make kind-install-cilium`. 

Now i see the shutdown logs (without an error) until the shutdown times-out after 1min. (Same behavior on `main`).